packages/grpc-js/src/client-interceptors.ts

```diff
@@ -0,0 +1,563 @@
+      onReceiveMetadata: metadata => {
+        interceptingListener?.onReceiveMetadata?.(metadata);
+      },
+      onReceiveMessage: message => {
+        // tslint:disable-next-line no-any
+        let deserialized: any;
+        try {
+          deserialized = this.methodDefinition.responseDeserialize(message);
+          interceptingListener?.onReceiveMessage?.(deserialized);
+        } catch (e) {
+          readError = {
+            code: Status.INTERNAL,
+            details: 'Failed to parse server response',
+            metadata: new Metadata(),
+          };
+          this.call.cancelWithStatus(readError.code, readError.details);
+        }
+      },
+      onReceiveStatus: status => {
+        if (readError) {
+          interceptingListener?.onReceiveStatus?.(readError);
+        } else {
+          interceptingListener?.onReceiveStatus?.(status);
+        }
+      },
+    });
+  }
+  startRead() {
+    this.call.startRead();
+  }
+  halfClose(): void {
+    this.call.halfClose();
+  }
+}
+
+/**
+ * BaseInterceptingCall with special-cased behavior for methods with unary
+ * responses.
+ */
+class BaseUnaryInterceptingCall extends BaseInterceptingCall
+  implements InterceptingCallInterface {
+  // tslint:disable-next-line no-any
+  constructor(call: Call, methodDefinition: ClientMethodDefinition<any, any>) {
+    super(call, methodDefinition);
+  }
+  start(metadata: Metadata, listener?: Partial<InterceptingListener>): void {
+    let receivedMessage = false;
+    const wrapperListener: InterceptingListener = {
+      onReceiveMetadata:
+        listener?.onReceiveMetadata?.bind(listener) ?? (metadata => {}),
+      // tslint:disable-next-line no-any
+      onReceiveMessage: (message: any) => {
+        receivedMessage = true;
+        listener?.onReceiveMessage?.(message);
+      },
+      onReceiveStatus: (status: StatusObject) => {
+        if (!receivedMessage) {
+          listener?.onReceiveMessage?.(null);
+        }
+        listener?.onReceiveStatus?.(status);
+      },
+    };
+    super.start(metadata, wrapperListener);
+    this.call.startRead();
+  }
+}
+
+/**
+ * BaseInterceptingCall with special-cased behavior for methods with streaming
+ * responses.
+ */
+class BaseStreamingInterceptingCall extends BaseInterceptingCall
+  implements InterceptingCallInterface {}
+
+// tslint:disable-next-line no-any
+function getBottomInterceptingCall(
+  channel: Channel,
+  path: string,
+  options: InterceptorOptions,
+  methodDefinition: ClientMethodDefinition<any, any>
+) {
+  const call = getCall(channel, path, options);
+  if (methodDefinition.responseStream) {
+    return new BaseStreamingInterceptingCall(call, methodDefinition);
+  } else {
+    return new BaseUnaryInterceptingCall(call, methodDefinition);
+  }
+}
+
+export interface NextCall {
+  (options: InterceptorOptions): InterceptingCallInterface;
+}
+
+export interface Interceptor {
+  (options: InterceptorOptions, nextCall: NextCall): InterceptingCall;
```

Comment:
`InterceptingCall` -> `InterceptingCallInterface`?

Reply:
I think we want it to be an actual `InterceptingCall` instance. That gives us some stronger guarantees about how the object was constructed, and about its implementation, which makes it more likely that it will work correctly. If you look at the example code in the gRFC I linked in the original post, you can see that even a do-nothing interceptor constructs a `new InterceptingCall` and returns it. I think enforcing that is a good thing.